Yosemite/Yosemite/Stores/ProductStore.swift

```diff
@@ -328,6 +330,22 @@ private extension ProductStore {
         }
     }
 
+    /// Retrieves the first product associated with a given siteID and exact-matching SKU (if any)
+    ///
+    func retrieveFirstProductMatchFromSKU(siteID: Int64, sku: String, onCompletion: @escaping (Result<Product, Error>) -> Void) {
+        remote.searchProductsBySKU(for: siteID, keyword: sku, pageNumber: Remote.Default.firstPageNumber, pageSize: 10, completion: { result in
```

Comment:
Are the results sorted somehow? I am wondering if setting the pageSize to 1 would give us a match, without having to query filter the array afterward.

Reply:
That's a good point, I reduced the pageSize to 1 in https://github.com/woocommerce/woocommerce-ios/pull/9698/commits/5abb31afcc99bde82d75bec2061a900049d58450 , however we still need to filter by `sku` afterward as otherwise we may get partial matches as first result (default behaviour as WC 6.6+). 

The tests `test_retrieveFirstProductMatchFromSKU_when_partial_SKU_match_then_returns_not_found_error` and `test_retrieveFirstProductMatchFromSKU_when_unsuccessful_SKU_match_then_returns_not_found_error` will fail as we can find partial matches in the simulated response (which we don't want).

Reply:
After further testing it seems we actually need a `pageSize` different that one, this is because of the partial match as default behaviour commented above: If we set the pageSize to 1, then the response from the API could be several partial matches until we finally get an exact match, which we will find by filtering the resulting array.

As is tricky to test with the current simulatedNetwork setup, I changed the pageSize to 1 and then manually tested for SKU matches in the current implementation of searching products:

| pageSize = 1 | pageSize = 25 |
|--------|--------|
| ![Simulator Screen Shot - iPhone 11 Pro - 2023-05-11 at 17 01 28](https://github.com/woocommerce/woocommerce-ios/assets/3812076/9c669658-17c5-4a21-bd05-08433fbb9559) | ![Simulator Screen Shot - iPhone 11 Pro - 2023-05-11 at 17 03 00](https://github.com/woocommerce/woocommerce-ios/assets/3812076/4f42270a-bfeb-4ae4-9099-50703fbc25d6) |

As the test shows, if we happen to limit the pageSize to 1 then  we would limit the products returned from the response, so we may be able to not find exact matches if there's enough partial matches. 

I reverted the change here and use the default items per page load (25) : 327e467fc53ca32e56499e63827a5cefdcb5b6ea

Reply:
I see you changed to 25, even with that, we can get 25 partial results and not the exact for this one right?

Reply:
>  we can get 25 partial results and not the exact for this one right?

Yes, that could be a case. An improvement I can think of is perhaps creating a new remote that is not the current `searchProductsBySKU`, which uses both API endpoints (partial-match and exact-match), but a new remote specifically for exact-matches only.

Reply:
Yes, or we could add a bigger value to the page size to be sure that it includes the exact match, but I guess 25 is big enough. Do we know if the endpoint returns the product by any specific order?

Reply:
> Yes, or we could add a bigger value to the page size to be sure that it includes the exact match, but I guess 25 is big enough. Do we know if the endpoint returns the product by any specific order?

It does, the [products endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) are returned ordered by `date_created`, so we get 25 recently created products by default.

We currently do not pass the `order_by` [parameter](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Networking/Networking/Remote/ProductsRemote.swift#L263-L269) when calling the API, but could if needed.